### PR TITLE
Add tracking for query rule types

### PR DESCRIPTION
**Pull request by kderusso**
This PR adds tracking for query rule types. 

The List Query Rulesets PR has been updated to include total rule types. Here is an example updated response: 
```
{
  "count": 2,
  "results": [
    {
      "ruleset_id": "my-ruleset",
      "rule_total_count": 3,
      "rule_criteria_types_counts": {
        "always": 3
      },
      "rule_type_counts": {
        "exclude": 1,
        "pinned": 2
      }
    },
    {
      "ruleset_id": "my-ruleset2",
      "rule_total_count": 3,
      "rule_criteria_types_counts": {
        "always": 3
      },
      "rule_type_counts": {
        "exclude": 1,
        "pinned": 2
      }
    }
  ]
}
```

The `_xpack/usage` call also propagates these stats. Here is an example of the updated `enterprise_search` section: 
```
"enterprise_search": {
    "available": true,
    "enabled": true,
    "search_applications": {
      "count": 0
    },
    "analytics_collections": {
      "count": 0
    },
    "query_rulesets": {
      "total_rule_count": 6,
      "rule_criteria_total_counts": {
        "always": 6
      },
      "total_count": 2,
      "min_rule_count": 3,
      "max_rule_count": 3,
      "rule_type_total_counts": {
        "pinned": 4,
        "exclude": 2
      }
    }
  }
```

### CODE REVIEW

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/rules/20_query_ruleset_list.yml [hunk 1]
```diff
@@ -307,3 +304,91 @@ teardown:
...
+        exact: 2
+        gt: 1
+        fuzzy: 1
+        prefix: 1
+        suffix: 1
+        always: 1
+  - match: { results.0.rule_type_counts: { pinned: 4, exclude: 1 } }
+
+
+
```
Comment (ioanatia) [inline x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/rules/20_query_ruleset_list.yml:394]: nitpicking - these lines are extra?

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/rules/20_query_ruleset_list.yml [hunk 2]
```diff
@@ -1,7 +1,4 @@
 setup:
-  - requires:
-      cluster_features: [ "gte_v8.10.0" ]
-      reason: Introduced in 8.10.0
```
Comment (ioanatia) [inline x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/rules/20_query_ruleset_list.yml:4]: is this going to be an issue when we backport to 8.x? not saying we should necessarily change it here, but maybe when we backport we might need to?

**Review by ioanatia — APPROVED** [review]

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/rules/20_query_ruleset_list.yml (hunk 1 above)
Comment (markjhoy) [reply x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/rules/20_query_ruleset_list.yml]: I removed them - looks like they were extra, but good to keep them clean.

**Comment by markjhoy** [comment]
buildkite test this

**Comment by markjhoy** [comment]
@ioanatia and @kderusso - I was able to get this merged - however, backporting this to the 8.16 and 8.15 branches are failing - and trying to run `backport --upstream elastic/elasticsearch --pr 116357` is showing a good amount of changes - especially in the `TransportVersions.java` file - which could be tricky if these changes are pushed there... any ideas / guidance?

**Comment by afoucret** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[](https://github.com/elastic/elasticsearch/pull/116820)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Pull request merged**